sdk-extension/opentelemetry-sdk-extension-aws/tests/trace/propagation/test_aws_xray_format.py

```diff
@@ -0,0 +1,375 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from requests.structures import CaseInsensitiveDict
+
+import opentelemetry.trace as trace_api
+from opentelemetry.sdk.extension.aws.trace.propagation.aws_xray_format import (
+    AwsXRayFormat,
+)
+from opentelemetry.trace import (
+    DEFAULT_TRACE_OPTIONS,
+    DEFAULT_TRACE_STATE,
+    INVALID_SPAN_CONTEXT,
+    SpanContext,
+    TraceFlags,
+    TraceState,
+    set_span_in_context,
+)
+
+TRACE_ID_BASE16 = "8a3c60f7d188f8fa79d48a391a778fa6"
+
+SPAN_ID_BASE16 = "53995c3f42cd8ad8"
+
+# Propagators Usage Methods
+
+
+def get_as_list(dict_object, key):
+    value = dict_object.get(key)
+    return [value] if value is not None else []
+
+
+# Inject Methods
+
+
+def build_test_context(
+    trace_id=int(TRACE_ID_BASE16, 16),
+    span_id=int(SPAN_ID_BASE16, 16),
+    is_remote=True,
+    trace_flags=DEFAULT_TRACE_OPTIONS,
+    trace_state=DEFAULT_TRACE_STATE,
+):
+    return set_span_in_context(
+        trace_api.DefaultSpan(
+            SpanContext(
+                trace_id, span_id, is_remote, trace_flags, trace_state,
+            )
+        )
+    )
+
+
+def build_dict_with_xray_trace_header(
+    trace_id="{}{}{}{}{}".format(
+        AwsXRayFormat.TRACE_ID_VERSION,
+        AwsXRayFormat.TRACE_ID_DELIMITER,
+        TRACE_ID_BASE16[: AwsXRayFormat.TRACE_ID_FIRST_PART_LENGTH],
+        AwsXRayFormat.TRACE_ID_DELIMITER,
+        TRACE_ID_BASE16[AwsXRayFormat.TRACE_ID_FIRST_PART_LENGTH :],
+    ),
+    span_id=SPAN_ID_BASE16,
+    sampled="0",
+):
+    carrier = CaseInsensitiveDict()
+
+    carrier[AwsXRayFormat.TRACE_HEADER_KEY] = "".join(
+        [
+            "{}{}{}{}".format(
```

Comment:
For tests I would write out the values instead of creating them with code. Makes much easier to read what the format is and there isn't really any maintenance overhead since they're perfect constants (i.e., format is defined by a spec not this code)

Reply:
I definitely see where you're coming from! I guess I thought with a string you can't really document the small changes that this test is testing like you can with this method where you have some default and then you _modify_ it to test a new edge case.

Specifically I was thinking you start with some base test like this:

`build_dict_with_xray_trace_header(),`

then you can test edge cases by indicating what part you're changing for this test like in the following examples:

`build_dict_with_xray_trace_header(sampled="1"),`

`build_dict_with_xray_trace_header(trace_id="abcdefghijklmnopqrstuvwxyz123456"),`

`build_dict_with_xray_trace_header(span_id="abcdefghijklmnop"),`

`build_dict_with_xray_trace_header(sampled=""),`

`build_dict_with_xray_trace_header(sampled="10002"),`

`build_dict_with_xray_trace_header(span_id="53995c3f42cd8ad800"),`

`build_dict_with_xray_trace_header(sampled="a"),`

I personally found this more reassuring than having to look at the string and see what is different about it. Although something like `Sampled=a` was the only thing that changed, I felt the need to check the long `Trace ID` and `Span ID`s in the strings to make sure they were the same. i.e.:

`"Root=1-8a3c60f7-d188f8fa79d48a391a778fa6;Parent=53995c3f42cd8ad8;Sampled=a"`

But I get that it's not _that_ long either. And once we get it merged the right way, you would be able to tell from the diff right away if someone tried to change `Trace ID` or `Span ID` in a PR. That way future readers can safely rely on it being done correct previously 😛 

I also concede that I "forced it to fit" when I did the following 3 complicated examples you commented on below like the following:

```
default_xray_trace_header_dict = build_dict_with_xray_trace_header()
xray_trace_header_dict_with_extra_fields = CaseInsensitiveDict(
    {
        AwsXRayFormat.TRACE_HEADER_KEY: default_xray_trace_header_dict[
            AwsXRayFormat.TRACE_HEADER_KEY
        ]
        + ";Foo=Bar"
    }
)
```

```
default_xray_trace_header_dict = build_dict_with_xray_trace_header()
trace_header_components = default_xray_trace_header_dict[
    AwsXRayFormat.TRACE_HEADER_KEY
].split(AwsXRayFormat.KV_PAIR_DELIMITER)
reversed_trace_header_components = AwsXRayFormat.KV_PAIR_DELIMITER.join(
    trace_header_components[::-1]
)
xray_trace_header_dict_in_different_order = CaseInsensitiveDict(
    {AwsXRayFormat.TRACE_HEADER_KEY: reversed_trace_header_components}
)
actual_context_encompassing_extracted = AwsXRayPropagatorTest.XRAY_PROPAGATOR.extract(
    AwsXRayPropagatorTest.carrier_getter,
    xray_trace_header_dict_in_different_order,
)
```

```
default_xray_trace_header_dict = build_dict_with_xray_trace_header()
trace_header_components = default_xray_trace_header_dict[
    AwsXRayFormat.TRACE_HEADER_KEY
].split(AwsXRayFormat.KV_PAIR_DELIMITER)
xray_trace_header_dict_with_extra_whitespace = CaseInsensitiveDict(
    {
        AwsXRayFormat.TRACE_HEADER_KEY: AwsXRayFormat.KV_PAIR_DELIMITER.join(
            [
                AwsXRayFormat.KEY_AND_VALUE_DELIMITER.join(
                    [
                        "     " + key + "     ",
                        "     " + value + "     ",
                    ]
                )
                for kv_pair_str in trace_header_components
                for key, value in [
                    kv_pair_str.split(
                        AwsXRayFormat.KEY_AND_VALUE_DELIMITER
                    )
                ]
            ]
        )
    }
)
```

I guess I would ask if you think it's worth it to keep the method for the small incremental changes like I mentioned above, and then write out the string _only_ for the complicated examples? Or maybe writing out the full string across the board is better for consistency?

Please let me know what you think! :)

Reply:
I would use the strings consistently - I think the test method names capture well what's special about the string without the helper.

Reply:
Sounds good, I've changed all of them to be dicts which contain the raw strings!